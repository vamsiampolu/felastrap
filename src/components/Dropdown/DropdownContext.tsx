import { createContext } from "react";
import { useInteractions } from "@floating-ui/react";

export type GetItemProps = ReturnType<typeof useInteractions>["getItemProps"];
export type HandleSelect = (index: number | null) => void;

export type DropdownContextValue = {
  activeIndex: number | null;
  selectedIndex: number | null;
  getItemProps: GetItemProps;
  handleSelect: HandleSelect;
};

export const DropdownContext = createContext({} as DropdownContextValue);
