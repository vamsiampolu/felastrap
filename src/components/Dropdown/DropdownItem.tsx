import { useListItem } from "@floating-ui/react";
import { useContext } from "react";
import { useFela } from "react-fela";
import type { StyleProps } from "../common";
import { DropdownContext } from "./DropdownContext";
export type DropDownItemProps = {
  label: string;
  disabled: boolean;
};

export type CssProps = {
  active: boolean;
  disabled: boolean;
};

const rule = () => {
  const color = "#212529";
  const backgroundColor = "transparent";

  return {
    backgroundColor,
    border: 0,
    borderRadius: 0,
    clear: "both",
    color,
    display: "block",
    fontWeight: 400,
    padding: "0.25rem 1rem",
    textAlign: "inherit",
    textDecoration: "none",
    whiteSpace: "nowrap",
    ":focus": {
      color: "#212529",
      backgroundColor: "#f8f9fa",
    },
    ":hover": {
      color: "#212529",
      backgroundColor: "#f8f9fa",
    },
  };
};

const activeRule = ({ active }: StyleProps<CssProps>) => {
  if (active) {
    return {
      backgroundColor: "#0d6efd",
      color: "#fff",
      textDecoration: "none",
    };
  }

  return {};
};

const disabledRule = ({ disabled }: StyleProps<CssProps>) => {
  if (disabled) {
    return {
      backgroundColor: "#ff",
      color: "#212529",
      opacity: "0.5",
      pointerEvents: "none",
    };
  }

  return {};
};

export const DropdownItem = ({
  label,
  disabled = false,
}: DropDownItemProps) => {
  const { activeIndex, selectedIndex, getItemProps, handleSelect } =
    useContext(DropdownContext);

  const { ref, index } = useListItem({ label });

  const isActive = activeIndex === index;
  const isSelected = selectedIndex === index;
  const tabIndex = isActive ? 0 : -1;
  const selected = isActive && isSelected;

  const onClick = () => handleSelect(index);
  const itemProps = getItemProps({ onClick });

  const { css } = useFela({ active: isActive, disabled });
  const className = `${css([rule, activeRule, disabledRule])} dropdown-item`;

  return (
    <li
      data-testid="dropdown-item"
      aria-disabled={disabled}
      aria-selected={selected}
      className={className}
      ref={ref}
      role="option"
      tabIndex={tabIndex}
      {...itemProps}
    >
      {label}
    </li>
  );
};
