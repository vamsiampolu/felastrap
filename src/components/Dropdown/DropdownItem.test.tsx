import { useListItem } from "@floating-ui/react";
import { render, screen } from "@testing-library/react";
import { IRenderer } from "fela";
import { RefObject, createRef } from "react";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../test-helpers";
import { DropdownContext, DropdownContextValue } from "./DropdownContext";
import { DropdownItem } from "./DropdownItem";

jest.mock("@floating-ui/react", () => ({
  ...jest.requireActual("@floating-ui/react"),
  useListItem: jest.fn(),
}));

describe("components/Dropdown/DropdownItem", () => {
  const getItemProps = () => ({});
  const handleSelect = () => undefined;
  const value = {
    activeIndex: -1,
    selectedIndex: -1,
    getItemProps,
    handleSelect,
  };

  let renderer: IRenderer;
  let ref: RefObject<unknown> | null;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);

    ref = createRef();
    (useListItem as jest.Mock).mockReturnValue({
      ref,
      index: 1,
    });
  });

  afterEach(() => {
    renderer.clear();

    ref = null;
    (useListItem as jest.Mock).mockClear();
  });

  it("should render a dropdown item", async () => {
    const component = (
      <DropdownContext.Provider value={value as DropdownContextValue}>
        <DropdownItem label="Dropdown Item" disabled={false} />
      </DropdownContext.Provider>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("dropdown-item")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-item")).toHaveClass("dropdown-item");

    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "role",
      "option",
    );
    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "tabIndex",
      "-1",
    );

    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "aria-disabled",
      "false",
    );
    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "aria-selected",
      "false",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
    ref = null;
  });

  it("should render a disabled dropdown item", async () => {
    const component = (
      <DropdownContext.Provider value={value as DropdownContextValue}>
        <DropdownItem label="Dropdown Item" disabled />
      </DropdownContext.Provider>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("dropdown-item")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-item")).toHaveClass("dropdown-item");

    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "aria-disabled",
      "true",
    );
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a selected dropdown item", async () => {
    const ctxValue: DropdownContextValue = {
      ...value,
      activeIndex: 1,
      selectedIndex: 1,
    };

    const component = (
      <DropdownContext.Provider value={ctxValue}>
        <DropdownItem label="Dropdown Item" disabled={false} />
      </DropdownContext.Provider>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);
    expect(screen.getByTestId("dropdown-item")).toBeInTheDocument();

    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "aria-selected",
      "true",
    );
    expect(screen.getByTestId("dropdown-item")).toHaveAttribute(
      "tabIndex",
      "0",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
