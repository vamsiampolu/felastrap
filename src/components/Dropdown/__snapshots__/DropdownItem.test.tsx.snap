// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Dropdown/DropdownItem should render a disabled dropdown item 1`] = `
".a {
  background-color: #ff;
}
.b {
  clear: both;
}
.c {
  color: #212529;
}
.d {
  display: block;
}
.e {
  font-weight: 400;
}
.f {
  text-align: inherit;
}
.g {
  white-space: nowrap;
}
.l {
  opacity: 0.5;
}
.m {
  pointer-events: none;
}
.n {
  text-decoration-line: none;
}
.o {
  padding-top: 0.25rem;
}
.p {
  padding-right: 1rem;
}
.q {
  padding-bottom: 0.25rem;
}
.r {
  padding-left: 1rem;
}
.s {
  border-top-left-radius: 0;
}
.t {
  border-top-right-radius: 0;
}
.u {
  border-bottom-right-radius: 0;
}
.v {
  border-bottom-left-radius: 0;
}
.w {
  border-top-width: 0;
}
.x {
  border-right-width: 0;
}
.y {
  border-bottom-width: 0;
}
.z {
  border-left-width: 0;
}
.j:hover {
  color: #212529;
}
.k:hover {
  background-color: #f8f9fa;
}
.h:focus {
  color: #212529;
}
.i:focus {
  background-color: #f8f9fa;
}


<li
  data-testid="dropdown-item"
  aria-disabled="true"
  aria-selected="false"
  class="a b c d e f g h i j k l m n o p q r s t u v w x y z dropdown-item"
  role="option"
  tabindex="-1"
>
  Dropdown Item
</li>
;
"
`;

exports[`components/Dropdown/DropdownItem should render a dropdown item 1`] = `
".a {
  background-color: transparent;
}
.b {
  clear: both;
}
.c {
  color: #212529;
}
.d {
  display: block;
}
.e {
  font-weight: 400;
}
.f {
  text-align: inherit;
}
.g {
  white-space: nowrap;
}
.l {
  text-decoration-line: none;
}
.m {
  padding-top: 0.25rem;
}
.n {
  padding-right: 1rem;
}
.o {
  padding-bottom: 0.25rem;
}
.p {
  padding-left: 1rem;
}
.q {
  border-top-left-radius: 0;
}
.r {
  border-top-right-radius: 0;
}
.s {
  border-bottom-right-radius: 0;
}
.t {
  border-bottom-left-radius: 0;
}
.u {
  border-top-width: 0;
}
.v {
  border-right-width: 0;
}
.w {
  border-bottom-width: 0;
}
.x {
  border-left-width: 0;
}
.j:hover {
  color: #212529;
}
.k:hover {
  background-color: #f8f9fa;
}
.h:focus {
  color: #212529;
}
.i:focus {
  background-color: #f8f9fa;
}


<li
  data-testid="dropdown-item"
  aria-disabled="false"
  aria-selected="false"
  class="a b c d e f g h i j k l m n o p q r s t u v w x dropdown-item"
  role="option"
  tabindex="-1"
>
  Dropdown Item
</li>
;
"
`;

exports[`components/Dropdown/DropdownItem should render a selected dropdown item 1`] = `
".a {
  background-color: #0d6efd;
}
.b {
  clear: both;
}
.c {
  color: #fff;
}
.d {
  display: block;
}
.e {
  font-weight: 400;
}
.f {
  text-align: inherit;
}
.g {
  white-space: nowrap;
}
.l {
  text-decoration-line: none;
}
.m {
  padding-top: 0.25rem;
}
.n {
  padding-right: 1rem;
}
.o {
  padding-bottom: 0.25rem;
}
.p {
  padding-left: 1rem;
}
.q {
  border-top-left-radius: 0;
}
.r {
  border-top-right-radius: 0;
}
.s {
  border-bottom-right-radius: 0;
}
.t {
  border-bottom-left-radius: 0;
}
.u {
  border-top-width: 0;
}
.v {
  border-right-width: 0;
}
.w {
  border-bottom-width: 0;
}
.x {
  border-left-width: 0;
}
.j:hover {
  color: #212529;
}
.k:hover {
  background-color: #f8f9fa;
}
.h:focus {
  color: #212529;
}
.i:focus {
  background-color: #f8f9fa;
}


<li
  data-testid="dropdown-item"
  aria-disabled="false"
  aria-selected="true"
  class="a b c d e f g h i j k l m n o p q r s t u v w x dropdown-item"
  role="option"
  tabindex="0"
>
  Dropdown Item
</li>
;
"
`;
